src/treelearner/serial_tree_learner.cpp

```diff
@@ -528,6 +540,159 @@ void SerialTreeLearner::FindBestSplitsFromHistograms(const std::vector<int8_t>&
   #endif
 }
 
+int32_t SerialTreeLearner::ForceSplits(Tree* tree, Json forced_split_json, int* left_leaf,
```

Comment:
I think the argument of these `Json` could be `const Json &`, since you didn't change their contents, and reference can save the copy time.